addon/ember-concurrency.d.ts

```diff
@@ -0,0 +1,139 @@
+declare module 'ember-concurrency' {
+  import EmberObject from '@ember/object';
+  import {
+    UnwrapComputedPropertyGetter,
+    UnwrapComputedPropertyGetters
+  } from '@ember/object/-private/types';
+
+  import RSVP from 'rsvp';
+
+  // Lifted from @types/ember__object/observable.d.ts
+  interface Getter {
+    /**
+     * Retrieves the value of a property from the object.
+     */
+    get<K extends keyof this>(key: K): UnwrapComputedPropertyGetter<this[K]>;
+    /**
+     * To get the values of multiple properties at once, call `getProperties`
+     * with a list of strings or an array:
+     */
+    getProperties<K extends keyof this>(
+      list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+    getProperties<K extends keyof this>(
+      ...list: K[]
+    ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+  }
+
+  export type GeneratorFn<Args extends any[] = any[], R = any> = (
+    ...args: Args
+  ) => IterableIterator<R>;
+
+  export const all: typeof Promise.all;
+  export const allSettled: typeof RSVP.allSettled;
+  export const hash: typeof RSVP.hash;
+  export const race: typeof Promise.race;
+
+  export function timeout(ms: number): Promise<void>;
+
+  export function waitForEvent(
+    object: EmberObject | EventTarget,
+    eventName: string
+  ): Promise<Event>;
+
+  export function waitForProperty<T extends object, K extends keyof T>(
+    object: T,
+    key: K,
+    callbackOrValue: (value: T[K]) => boolean
+  ): Promise<void>;
+
+  export function waitForQueue(queueName: string): Promise<void>;
+
+  export function task<Args extends any[], R>(
+    taskFn: GeneratorFn<Args, R>
+  ): Task<Args, Exclude<R, Promise<any>>>;
+  export function task<Args extends any[], R>(encapsulatedTask: {
+    perform: GeneratorFn<Args, R>;
+  }): Task<Args, Exclude<R, Promise<any>>>;
+
+  export function taskGroup(): TaskGroupProperty;
+
+  interface CommonTaskProperty {
+    restartable: () => TaskProperty;
+    drop: () => TaskProperty;
+    keepLatest: () => TaskProperty;
+    enqueue: () => TaskProperty;
+    maxConcurrency: (n: number) => TaskProperty;
+    cancelOn: (eventName: string) => TaskProperty;
+    group: (groupName: string) => TaskProperty;
+  }
+
+  export interface TaskProperty extends CommonTaskProperty {
+    evented: () => TaskProperty;
+    debug: () => TaskProperty;
+    on: (eventName: string) => TaskProperty;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-interface
+  export interface TaskGroupProperty extends CommonTaskProperty {}
+
+  // Based on https://github.com/CenterForOpenScience/ember-osf-web/blob/7933316efae805e00723789809bdeb58a96a286a/types/ember-concurrency/index.d.ts
+
+  export enum TaskInstanceState {
+    Dropped = 'dropped',
+    Canceled = 'canceled',
+    Finished = 'finished',
+    Running = 'running',
+    Waiting = 'waiting'
+  }
+
+  export interface TaskInstance<T> extends PromiseLike<T>, Getter {
```

Comment:
We may want to tweak this to be a union type where `TaskInstanceState` and the various booleans are all always guaranteed/required to be the same.

```ts
interface TaskInstanceBase<T> extends PromiseLike<T>, Getter {
    readonly hasStarted: boolean;
    readonly isCanceled: boolean;
    readonly isDropped: boolean;
    readonly isFinished: boolean;
    readonly isRunning: boolean;
    readonly isSuccessful: boolean;
    readonly value?: T;
    cancel(): void;
    catch(): RSVP.Promise<unknown>;
    finally(): RSVP.Promise<unknown>;
    then<TResult1 = T, TResult2 = never>(
      onfulfilled?:
        | ((value: T) => TResult1 | RSVP.Promise<TResult1>)
        | undefined
        | null,
      onrejected?:
        | ((reason: any) => TResult2 | PromiseLike<TResult2>)
        | undefined
        | null
    ): RSVP.Promise<TResult1 | TResult2>;
}

interface InstanceDropped {
    readonly state: TaskInstanceState.Dropped;
    hasStarted: true; // ?
    isCanceled: true; // ?
    isDropped: true;
    isError: false; // ?
    isFinished: true; // ?
    isRunning: false; // ?
    isSuccessful: false; // ?
}

interface InstanceCanceled {
    readonly state: TaskInstanceState.Canceled;
    hasStarted: true; // ?
    isCanceled: true;
    isDropped: false; // ?
    isError: false; // ?
    isFinished: true; // ?
    isRunning: false; // ?
    isSuccessful: false; // ?
};

// etc.

interface InstanceError {
    isError: true;
    error: unknown;
    hasStarted: true; // ?
    isCanceled: false; // ?
    isDropped: false; // ?
    isFinished: true;
    isRunning: false;
    isSuccessful: false;
}

interface Success {
    isError: false;
    error: undefined;
    hasStarted: true; // ?
    isCanceled: false; // ?
    isDropped: false; // ?
    isError: false;
    isFinished: true;
    isRunning: false;
    isSuccessful: true;
}

export type TaskInstance<T> = TaskInstanceBase<T> & (
    | InstanceDropped
    | InstanceCanceled
    | InstanceFinished
    | InstanceRunning
    | InstanceWaiting
    | InstanceError
    | InstanceSuccess
);
```

The `// ?`s are for all the places where I don't remember from Ember Concurrency's docs what the actual behavior is – some of those may be `boolean` *always* instead of a `const` type like `true` or `false`.